### DISCUSSION TITLE
How can we test in continuous the semantic layer?

### DISCUSSION THREAD

**Discussion by mchl-schrdng** [Canner/WrenAI#372] [Ideas]
Utilizing a semantic layer for large language model (LLM) usage is highly beneficial. However, what if there's an error in the database? What if the schema is not synchronized with the semantic definition? One straightforward solution could be to employ the semantic layer to execute a generic SQL statement with a limit of zero, thereby incurring no cost. This approach would help identify any errors in the layer. In other tools like Looker, this test is run during pull requests or on a daily basis to detect any mismatches between the semantic layer and the underlying database. It would be advantageous to have this feature integrated into the LLM system as well.

**Comment by wwwy3y3** [comment]
Hi @mchl-schrdng thanks for opening this discussion and sharing your opinions with us!

> Utilizing a semantic layer for large language model (LLM) usage is highly beneficial. However, what if there's an error in the database? What if the schema is not synchronized with the semantic definition? One straightforward solution could be to employ the semantic layer to execute a generic SQL statement with a limit of zero, thereby incurring no cost. This approach would help identify any errors in the layer. 

We're indeed working on a feature that would scan schema conflicts between semantic definitions and the data source. The scan could be triggered manually by user or by background process on daily basis. We're working on it in this PR: https://github.com/Canner/WrenAI/pull/339. 

The idea is to check if any tables/columns used in semantic definitions got deleted or altered and show warnings on our UI.

> In other tools like Looker, this test is run during pull requests or on a daily basis to detect any mismatches between the semantic layer and the underlying database.

Didn't know Looker does that. Thanks for sharing!

Feel free to share more with us.

**Reply by mchl-schrdng** [reply]
Oh, really great.

Yes, on Looker you can use an open-source solution ([Spectacles](https://github.com/spectacles-ci/spectacles)), where you can run two kinds of tests:

Default Mode
- Purpose: Find all errors by querying each dimension.
- Operation: Runs one query per Explore. If an error is found, runs individual queries for each dimension to identify specific errors.
- Pros: Identifies all errors in one run.
- Cons: More queries, longer runtime.

Fail-Fast Mode
- Purpose: Speed up validation by running fewer queries.
- Operation: Runs one query per Explore without querying each dimension.
- Pros: Faster, fewer queries.
- Cons: Only the first error in each Explore is found, potentially requiring multiple runs to catch all errors.

Both checks are performed with a limit of 0 to avoid any cost.

**Reply by wwwy3y3** [reply]
Really detailed sharing! thanks @mchl-schrdng. We'll see if we could get some inspirations from [Spectacles](https://github.com/spectacles-ci/spectacles)

**Reply by wwwy3y3** [reply]
For data warehouse like bigquery, even if you run limit 0 or 1, I think there's still cost behind that. https://stackoverflow.com/a/63174099/1191953

We're thinking of running statements like `describe` to validate if the statements are valid.

**Reply by mchl-schrdng** [reply]
With a limit 1, yes you are still scanning.
But with limit 0, you don't (cf screenshot).
<img width="252" alt="image" src="https://github.com/Canner/WrenAI/assets/73759636/e298a6f6-e825-409c-a41a-012df3975ef8">

**Reply by wwwy3y3** [reply]
@mchl-schrdng I see. Then, it might be a good way to validate the statement! thanks for the screenshot!

**Reply by wwwy3y3** [reply]
cc @cyyeh @goldmedal @onlyjackfrost
